In [39]:
import pandas as pd

In [40]:
import networkx as nx
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:

network = nx.DiGraph()

# Load nodes
nodes_df = pd.read_csv('Nodes.csv')
for index, row in nodes_df.iterrows():
    network.add_node(row['Id'], **row.to_dict())

# Load edges
edges_df = pd.read_csv('Edges.csv', delimiter=';')
for index, row in edges_df.iterrows():
    network.add_edge(row['Source'], row['Target'])

In [55]:
nodes_df

,Unnamed: 0,Id,en_curid,Name,NameURL,Link,birthcity,countryName,countryCode_alpha2,countryCode_alpha3,...,harmonicClosenessCentrality,betweenessCentrality,authority,hub,indegree,outdegree,degree,clustering,eigenCentrality,BCI
0,1,0,269846,Alexander I of Yugoslavia,Alexander_I_of_Yugoslavia,https://en.wikipedia.org/wiki/Alexander_I_of_Y...,CETINJE,NaN,ME,MNE,...,0.231996,1.980177e+10,0.000060,0.000068,10,8,18,0.090476,0.010304,0.004918
1,2,1,585020,Aung San,Aung_San,https://en.wikipedia.org/wiki/Aung_San,NATMAUK,NaN,MM,MMR,...,0.229009,2.013823e+08,0.000023,0.000071,4,6,10,0.285714,0.003036,0.001816
2,3,2,2847,Aung San Suu Kyi,Aung_San_Suu_Kyi,https://en.wikipedia.org/wiki/Aung_San_Suu_Kyi,YANGON,NaN,MM,MMR,...,0.265566,2.402678e+11,0.000344,0.000722,24,29,53,0.045296,0.032603,0.022338
3,4,3,89446,Chow Yun-Fat,Chow_Yun-Fat,https://en.wikipedia.org/wiki/Chow_Yun-Fat,LAMMA ISLAND,NaN,HK,HKG,...,0.228340,5.783984e+09,0.000038,0.000076,2,6,8,0.100000,0.004640,0.001853
4,5,4,1421152,Claude Makélélé,Claude_Makélélé,https://en.wikipedia.org/wiki/Claude_Makélélé,KINSHASA,NaN,CD,COD,...,0.246165,4.006155e+10,0.000580,0.000867,24,27,51,0.125436,0.063908,0.009841
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11335,11336,11335,2571986,Cara Black,Cara_Black,https://en.wikipedia.org/wiki/Cara_Black,HARARE,ZIMBABWE,ZW,ZWE,...,0.208072,8.972075e+08,0.040806,0.022348,23,14,37,0.556923,0.099160,0.003015
11336,11337,11336,205137,Ian Smith,Ian_Smith,https://en.wikipedia.org/wiki/Ian_Smith,SHURUGWI,ZIMBABWE,ZW,ZWE,...,0.248704,2.076401e+10,0.000171,0.000530,13,14,27,0.166667,0.034723,0.007752
11337,11338,11337,915950,Kirsty Coventry,Kirsty_Coventry,https://en.wikipedia.org/wiki/Kirsty_Coventry,HARARE,ZIMBABWE,ZW,ZWE,...,0.204955,6.944703e+09,0.000000,0.000008,2,2,4,0.000000,0.000731,0.004010
11338,11339,11338,21492584,Morgan Tsvangirai,Morgan_Tsvangirai,https://en.wikipedia.org/wiki/Morgan_Tsvangirai,GUTU,ZIMBABWE,ZW,ZWE,...,0.223834,3.098468e+09,0.000075,0.000156,7,6,13,0.136364,0.007059,0.004778


In [56]:
edges_df

,Source,Target,Id,linkname
0,0,316,0.0,0->316
1,0,625,1.0,0->625
2,0,2312,2.0,0->2312
3,0,4153,3.0,0->4153
4,0,6158,4.0,0->6158
...,...,...,...,...
126273,3,33,NaN,7669->7958
126274,6,1,NaN,9900->0
126275,6,41,NaN,9900->9171
126276,6,38,NaN,9900->8651


In [42]:
year = [-200000, 0]
target = nodes_df[(nodes_df['birthyear'] < year[1]) & (nodes_df['birthyear'] > year[0])]
target

,Unnamed: 0,Id,en_curid,Name,NameURL,Link,birthcity,countryName,countryCode_alpha2,countryCode_alpha3,...,harmonicClosenessCentrality,betweenessCentrality,authority,hub,indegree,outdegree,degree,clustering,eigenCentrality,BCI
37,38,37,262156,Roxana,Roxana,https://en.wikipedia.org/wiki/Roxana,OTHER,AFGHANISTAN,AF,AFG,...,0.188776,2.100000e+00,0.000006,0.000016,6,5,11,0.690476,0.013360,0.007569
38,39,38,44406,Zoroaster,Zoroaster,https://en.wikipedia.org/wiki/Zoroaster,AIRYANEM VAEJAH,AFGHANISTAN,AF,AFG,...,0.266749,1.358638e+11,0.000019,0.000197,12,39,51,0.054965,0.013790,0.016847
508,509,508,1698,Ambiorix,Ambiorix,https://en.wikipedia.org/wiki/Ambiorix,OTHER,BELGIUM,BE,BEL,...,0.195133,0.000000e+00,0.000003,0.000010,2,1,3,0.666667,0.003947,0.002126
799,800,799,222779,Diogenes Apolloniates,Diogenes_Apolloniates,https://en.wikipedia.org/wiki/Diogenes_Apollon...,SOZOPOL,BULGARIA,BG,BGR,...,0.211351,2.928585e+06,0.000004,0.000023,1,5,6,0.300000,0.000683,0.000967
805,806,805,18993992,Spartacus,Spartacus,https://en.wikipedia.org/wiki/Spartacus,THRACE,BULGARIA,BG,BGR,...,0.224073,4.063992e+10,0.000113,0.000050,12,9,21,0.068627,0.025846,0.013709
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11252,11253,11252,608209,Xun Zi,Xun_Zi,https://en.wikipedia.org/wiki/Xun_Zi,OTHER,UNKNOWN,UNK,UNK,...,0.180885,1.567713e+10,0.000001,0.000002,3,6,9,0.309524,0.002835,0.003845
11257,11258,11257,68823,Zedekiah,Zedekiah,https://en.wikipedia.org/wiki/Zedekiah,OTHER,UNKNOWN,UNK,UNK,...,0.179763,1.709096e+09,0.000003,0.000001,5,4,9,0.366667,0.003622,0.003845
11322,11323,11322,405124,Hud,Hud_(prophet),https://en.wikipedia.org/wiki/Hud_(prophet),æÀ€ÛD,YEMEN,YE,YEM,...,0.174741,0.000000e+00,0.000000,0.000000,0,1,1,0.000000,0.000000,0.000795
11324,11325,11324,357200,Queen of Sheba,Queen_of_Sheba,https://en.wikipedia.org/wiki/Queen_of_Sheba,MARIB,YEMEN,YE,YEM,...,0.243724,4.126094e+09,0.000023,0.000052,7,17,24,0.019763,0.005961,0.005807


In [43]:
year = [1840, 1945]
target = nodes_df[(nodes_df['deathyear'] < year[1]) & (nodes_df['deathyear'] > year[0])]
target

,Unnamed: 0,Id,en_curid,Name,NameURL,Link,birthcity,countryName,countryCode_alpha2,countryCode_alpha3,...,harmonicClosenessCentrality,betweenessCentrality,authority,hub,indegree,outdegree,degree,clustering,eigenCentrality,BCI
0,1,0,269846,Alexander I of Yugoslavia,Alexander_I_of_Yugoslavia,https://en.wikipedia.org/wiki/Alexander_I_of_Y...,CETINJE,NaN,ME,MNE,...,0.231996,1.980177e+10,0.000060,0.000068,10,8,18,0.090476,0.010304,0.004918
20,21,20,233350,Niels Ryberg Finsen,Niels_Ryberg_Finsen,https://en.wikipedia.org/wiki/Niels_Ryberg_Finsen,TÌÒRSHAVN,NaN,FO,FRO,...,0.000000,0.000000e+00,0.000000,0.000000,0,0,0,0.000000,0.000000,0.001128
26,27,26,37399,Roland Garros,Roland_Georges_Garros,https://en.wikipedia.org/wiki/Roland_Georges_G...,SAINT-DENIS,NaN,RE,NaN,...,0.000000,0.000000e+00,0.000000,0.000000,0,0,0,0.000000,0.000000,0.000820
33,34,33,2695,Abdur Rahman Khan,Abdur_Rahman_Khan,https://en.wikipedia.org/wiki/Abdur_Rahman_Khan,KABUL,AFGHANISTAN,AF,AFG,...,0.202263,2.289081e+06,0.000000,0.000022,1,3,4,0.083333,0.000494,0.001095
43,44,43,314537,Dost Mohammad Khan,Dost_Mohammad_Khan,https://en.wikipedia.org/wiki/Dost_Mohammad_Khan,KANDAHAR,AFGHANISTAN,AF,AFG,...,0.000000,0.000000e+00,0.000001,0.000000,4,0,4,0.083333,0.001578,0.001977
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10786,10787,10786,89182,Levi P. Morton,Levi_P._Morton,https://en.wikipedia.org/wiki/Levi_P._Morton,SHOREHAM,UNITED STATES,US,USA,...,0.222880,4.388805e+09,0.000063,0.000109,9,12,21,0.238095,0.018440,0.008070
10819,10820,10819,162385,Thorstein Veblen,Thorstein_Veblen,https://en.wikipedia.org/wiki/Thorstein_Veblen,CATO,UNITED STATES,US,USA,...,0.224658,3.809899e+09,0.000034,0.000051,6,6,12,0.081818,0.005831,0.002691
11263,11264,11263,154771,Comte de Lautréamont,Comte_de_Lautréamont,https://en.wikipedia.org/wiki/Comte_de_Lautréa...,MONTEVIDEO,URUGUAY,UY,URY,...,0.265028,8.290241e+09,0.000029,0.000191,3,30,33,0.059476,0.004408,0.001911
11272,11273,11272,315831,José Gervasio Artigas,José_Gervasio_Artigas,https://en.wikipedia.org/wiki/José_Gervasio_Ar...,MONTEVIDEO,URUGUAY,UY,URY,...,0.191573,4.892705e+09,0.000007,0.000001,3,2,5,0.250000,0.003543,0.001794


In [44]:
nodes_df[nodes_df['Link'] == 'https://en.wikipedia.org/wiki/Diogenes']

,Unnamed: 0,Id,en_curid,Name,NameURL,Link,birthcity,countryName,countryCode_alpha2,countryCode_alpha3,...,harmonicClosenessCentrality,betweenessCentrality,authority,hub,indegree,outdegree,degree,clustering,eigenCentrality,BCI


In [45]:
nodes_df[nodes_df['Name'] == 'Abbas']

,Unnamed: 0,Id,en_curid,Name,NameURL,Link,birthcity,countryName,countryCode_alpha2,countryCode_alpha3,...,harmonicClosenessCentrality,betweenessCentrality,authority,hub,indegree,outdegree,degree,clustering,eigenCentrality,BCI


In [46]:
nodes_df[nodes_df['Name'] == 'Ahmad Shah Massoud']

,Unnamed: 0,Id,en_curid,Name,NameURL,Link,birthcity,countryName,countryCode_alpha2,countryCode_alpha3,...,harmonicClosenessCentrality,betweenessCentrality,authority,hub,indegree,outdegree,degree,clustering,eigenCentrality,BCI


In [47]:
nodes_df[nodes_df['Id'] == 1405]

,Unnamed: 0,Id,en_curid,Name,NameURL,Link,birthcity,countryName,countryCode_alpha2,countryCode_alpha3,...,harmonicClosenessCentrality,betweenessCentrality,authority,hub,indegree,outdegree,degree,clustering,eigenCentrality,BCI
1405,1406,1405,58726,Margaret I of Denmark,Margaret_I_of_Denmark,https://en.wikipedia.org/wiki/Margaret_I_of_De...,VORDINGBORG,DENMARK,DK,DNK,...,0.215122,3.847766e+10,0.000007,0.000005,6,8,14,0.063636,0.003482,0.002913


In [48]:
nodes_df.index[nodes_df['NameURL'].isna()]

Index([], dtype='int64')

In [49]:
len(network.nodes)

22460

In [54]:
print(f"Network has {network.number_of_nodes()} nodes and {network.number_of_edges()} edges.")

Network has 22460 nodes and 126276 edges.


In [50]:
from pathlib import Path

folder_path = Path("wikipages")

file_count = sum(1 for item in folder_path.iterdir() if item.is_file())

print(f"The folder contains {file_count} files.")

The folder contains 11324 files.


In [72]:
print(len(edges_df[edges_df['Source'] == "2024"]))
print(len(edges_df[edges_df['Source'] == "2025"]))
print(len(edges_df[edges_df['Target'] == "2024"]))
print(len(edges_df[edges_df['Target'] == "2025"]))

0
15
0
33


In [73]:
print(len(edges_df[edges_df['Source'] == "4987"]))
print(len(edges_df[edges_df['Source'] == "8038"]))
print(len(edges_df[edges_df['Target'] == "4987"]))
print(len(edges_df[edges_df['Target'] == "8038"]))

0
36
0
41


In [76]:
print(len(edges_df[edges_df['Source'] == "7669"]))
print(len(edges_df[edges_df['Source'] == "8234"]))
print(len(edges_df[edges_df['Target'] == "7669"]))
print(len(edges_df[edges_df['Target'] == "8234"]))

0
9
0
14


In [75]:
n1 = "7921"
n2 = "9842"
print(len(edges_df[edges_df['Source'] == n1]))
print(len(edges_df[edges_df['Source'] == n2]))
print(len(edges_df[edges_df['Target'] == n1]))
print(len(edges_df[edges_df['Target'] == n2]))

0
8
0
7


In [96]:
n1 = "7656"
n2 = "9880"
print(len(edges_df[edges_df['Source'] == n1]))
print(len(edges_df[edges_df['Source'] == n2]))
print(len(edges_df[edges_df['Target'] == n1]))
print(len(edges_df[edges_df['Target'] == n2]))

5
5
6
6


In [97]:
edges_df[edges_df['Source'] == n1]

,Source,Target,Id,linkname
75561,7656,3814,75630.0,7656->3814
75562,7656,3847,75631.0,7656->3847
75563,7656,6256,75632.0,7656->6256
75564,7656,8153,75633.0,7656->8153
75565,7656,8283,75634.0,7656->8283


In [98]:
edges_df[edges_df['Source'] == n2]

,Source,Target,Id,linkname
108003,9880,3814,108142.0,9880->3814
108004,9880,3847,108143.0,9880->3847
108005,9880,6256,108144.0,9880->6256
108006,9880,8153,108145.0,9880->8153
108007,9880,8283,108146.0,9880->8283


9880->3814

9880->3847

9880->6256

9880->8153

9880->8283

In [107]:
network.nodes[3814]

{'Unnamed: 0': 3815,
 'Id': 3814,
 'en_curid': 97509,
 'Name': 'Michael Collins',
 'NameURL': 'Michael_Collins_(Irish_leader)',
 'Link': 'https://en.wikipedia.org/wiki/Michael_Collins_(Irish_leader)',
 'birthcity': 'CLONAKILTY',
 'countryName': 'IRELAND',
 'countryCode_alpha2': 'IE',
 'countryCode_alpha3': 'IRL',
 'longitude': -8.89,
 'latitude': 51.62,
 'continentName': 'EUROPE',
 'birthyear': 1890.0,
 'deathyear': 1922.0,
 'agespan': 32.0,
 'gender': 'Male',
 'occupation': 'POLITICIAN',
 'industry': 'GOVERNMENT',
 'domain': 'INSTITUTIONS',
 'domain2': 'GOVERNMENT',
 'L_star': 11.4783988,
 'numLangs': 47,
 'averageViews': 7592.468085,
 'totalPageViews': 356846,
 'stdDevPageViews': 6130.494528,
 'pageViewsEnglish': 140287,
 'pageViewsNonEnglish': 216559,
 'HPI': 21.83485284,
 'timeInterval': '<[1890.0, 1922.0]>',
 'pageRank': 0.000105,
 'eccentricity': 9,
 'closenessCentrality': 0.235766,
 'harmonicClosenessCentrality': 0.250711,
 'betweenessCentrality': 36738684618.0,
 'authority': 9.

In [99]:
edges_df[edges_df['Target'] == n1]

,Source,Target,Id,linkname
26145,2551,7656,26184.0,2551->7656
37155,3654,7656,37197.0,3654->7656
38494,3814,7656,38536.0,3814->7656
38617,3822,7656,38659.0,3822->7656
84847,8277,7656,84958.0,8277->7656
114878,10287,7656,115034.0,10287->7656


In [100]:
edges_df[edges_df['Target'] == n2]

,Source,Target,Id,linkname
26146,2551,9880,26185.0,2551->9880
37156,3654,9880,37198.0,3654->9880
38498,3814,9880,38540.0,3814->9880
38624,3822,9880,38666.0,3822->9880
84851,8277,9880,84962.0,8277->9880
114884,10287,9880,115040.0,10287->9880


2551->9880 (James Connolly UK) opposite
3654->9880 true
3814->9880 opposite
3822->9880 true
8277->9880 opposite
10287->9880 true

In [106]:
network.nodes[3654]

{'Unnamed: 0': 3655,
 'Id': 3654,
 'en_curid': 277700,
 'Name': 'V. V. Giri',
 'NameURL': 'V._V._Giri',
 'Link': 'https://en.wikipedia.org/wiki/V._V._Giri',
 'birthcity': 'BERHAMPUR',
 'countryName': 'INDIA',
 'countryCode_alpha2': 'IN',
 'countryCode_alpha3': 'IND',
 'longitude': 84.78,
 'latitude': 19.32,
 'continentName': 'ASIA',
 'birthyear': 1894.0,
 'deathyear': 1980.0,
 'agespan': 86.0,
 'gender': 'Male',
 'occupation': 'POLITICIAN',
 'industry': 'GOVERNMENT',
 'domain': 'INSTITUTIONS',
 'domain2': 'GOVERNMENT',
 'L_star': 2.603225691,
 'numLangs': 29,
 'averageViews': 13537.10345,
 'totalPageViews': 392576,
 'stdDevPageViews': 3983.081801,
 'pageViewsEnglish': 325436,
 'pageViewsNonEnglish': 67140,
 'HPI': 19.1999995,
 'timeInterval': '<[1894.0, 1980.0]>',
 'pageRank': 2.1e-05,
 'eccentricity': 10,
 'closenessCentrality': 0.21309,
 'harmonicClosenessCentrality': 0.225243,
 'betweenessCentrality': 776836758.0,
 'authority': 1.4e-05,
 'hub': 2.4e-05,
 'indegree': 1,
 'outdegree':

In [83]:
n1 = "8716"
n2 = "9900"
print(len(edges_df[edges_df['Source'] == n1]))
print(len(edges_df[edges_df['Source'] == n2]))
print(len(edges_df[edges_df['Target'] == n1]))
print(len(edges_df[edges_df['Target'] == n2]))

8
0
13
0


In [84]:
n1 = "7888"
n2 = "10099"
print(len(edges_df[edges_df['Source'] == n1]))
print(len(edges_df[edges_df['Source'] == n2]))
print(len(edges_df[edges_df['Target'] == n1]))
print(len(edges_df[edges_df['Target'] == n2]))

0
20
0
28


In [110]:
import io
import networkx as nx
import urllib.request

url = "https://raw.githubusercontent.com/denisababeii/socialgraphs_project/refs/heads/main/celebrities_clean_without_content.graphml"

# Download the content of the raw file from Github with the network information
with urllib.request.urlopen(url) as response:
    data = response.read()

# Load directly into NetworkX
network = nx.read_graphml(io.BytesIO(data))

In [111]:
len(network.nodes)

11286